### PR TITLE
M2 - implement consensus checkpointing algorithms 

### PR DESCRIPTION
**Pull request by polarbeargo**
**Summary of changes**
Changes introduced in this pull request:
- 
Base on the branch [noot/checkpoint](https://github.com/ChainSafe/mina-rs/commits/noot/checkpoint) continue with issue #24 

- [x] Add `initCheckpoints`, and `isShortRange` algorithms outlined here: https://github.com/MinaProtocol/mina/blob/32f529d8d8d712a44ee75be66061ce08cbdc8924/docs/specs/consensus/README.md#53-decentralized-checkpointing
- [x] Add defined `consensus_state` type.
- [x] Add license in `checkpoint.rs`.
- [x] Update library.
- [x] Fix: Use of moved value.
- [x] Feat: Implement `is_short_range()` for short-range fork check, `test_init_checkpoints` and , `init_checkpoints` for  genesis checkpoints follow the new spec.
- [x] Test: Implement `test_is_short_range()` - Add both true and false test cases.
- [x]  Fix: Replace Common to Chain trait.
- [x]  chore: Recover Cargo.lock.

**Reference issue to close (#24 )**

Closes Closes 
Cargo build & Cargo test passed.
Screenshot (291)
**Specification reference**

- 
https://github.com/MinaProtocol/mina/blob/32f529d8d8d712a44ee75be66061ce08cbdc8924/docs/specs/consensus/README.md

**Other information and links**

https://github.com/MinaProtocol/mina/blob/develop/src/lib/consensus/proof_of_stake.ml

### CODE REVIEW

File: consensus/src/common.rs [hunk 1]
```diff
@@ -30,7 +30,48 @@ where
     fn state_hash(&self) -> Option<StateHash>;
 }
 
-impl Chain<ProtocolState> for ProtocolStateChain {
+impl Common<ProtocolState> for ProtocolState {
```
Comment (creativcoder) [inline consensus/src/common.rs:33]: This should be reverted back. Can you explain why we had to re-implement this. These methods are common across all ProtocolState instances.

File: consensus/src/checkpoint.rs [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+        );
+        assert_eq!(
+            genesis.body.consensus_state.next_epoch_data.lock_checkpoint,
+            Base58Encodable::from_base58("3NLoKn22eMnyQ7rxh5pxB6vBA3XhSAhhrf7akdqS6HbAKD14Dh1d")
+                .unwrap()
+        );
+    }
+
+    #[wasm_bindgen_test]
+    fn test_is_short_range() {
```
Comment (creativcoder) [inline consensus/src/checkpoint.rs:120]: This only tests the `true` path. Can we add a test which asserts for alternative test case? Also we can look into proof_of_stake.ml and see if we can include those tests as well.

File: consensus/src/checkpoint.rs [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+
+use crate::common::{Common, ProtocolStateChain};
+use mina_crypto::base58::Base58Encodable;
+use mina_crypto::hash::{EpochSeed, StateHash};
+use mina_rs_base::protocol_state::ProtocolState;
+use wasm_bindgen_test::*;
+const SLOTS_PER_EPOCH: u32 = 7140;
+
+/// init_checkpoints initializes the checkpoints for the genesis block
+/// This function assumes the state hash of `genesis` is already set
```
Comment (creativcoder) [inline consensus/src/checkpoint.rs:12]: can remove this comment, as we do initialize the two checkpoints within this function.

File: consensus/src/common.rs (hunk 1 above)
Comment (polarbeargo) [reply consensus/src/common.rs:33]: This is from @noot's legacy code as the following:
Screenshot (285)
Screenshot (286)
We have both ProtocolStateChain and ProtocolState.

Comment (polarbeargo) [reply consensus/src/common.rs:33]: This part is from @noot's legacy code
Screenshot (285)
Screenshot (286)
:

File: consensus/src/checkpoint.rs (hunk 2 above)
Comment (polarbeargo) [reply consensus/src/checkpoint.rs:120]: Will do. I will skip the selection chain test case part because we haven't reached the selection part.

File: consensus/src/common.rs (hunk 1 above)
Comment (creativcoder) [reply consensus/src/common.rs:33]: As in https://github.com/ChainSafe/mina-rs/pull/34/files#diff-09b1b8c7ec19c8817ec99388da62c7d423b51a868df0ec116ef9d4dde9919286R21 the `Chain` trait should be left as it is. `ProtocolStateChain` only implements that. We would just need to update the methods within that trait to match the new spec, if there are any. If you see this section: https://github.com/MinaProtocol/mina/blob/32f529d8d8d712a44ee75be66061ce08cbdc8924/docs/specs/consensus/README.md#51-common those are all the methods the Chain trait implements. Also we don't need to strictly abide by the exact names. The spec naming convention is just for readability purposes. We can rename variables and type names as it makes sense in our code. So that's why we don't have it as `Common` but rather as `Chain`

Comment (polarbeargo) [reply consensus/src/common.rs:33]: Got it! Thank you for clarifying, Rahul.

File: consensus/src/common.rs [hunk 4]
```diff
@@ -30,6 +30,47 @@ where
     fn state_hash(&self) -> Option<StateHash>;
 }
 
+impl Chain<ProtocolState> for ProtocolState {
```
Comment (creativcoder) [inline consensus/src/common.rs:33]: I don't see why we need to implement Chain here on ProtocolState, as it's already implemented on `ProtocolStateChain`. Also these API's only make sense on a blockchain. ProtocolState is a just a single block header, so by definition it should not be implementing Chain trait.

Comment (polarbeargo) [reply consensus/src/common.rs:33]: Ok, I will remove `ProtocolState` from  `Chain `trait.

File: consensus/src/checkpoint.rs [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+            .unwrap()
+            .staking_epoch_data
+            .lock_checkpoint
+            == c1
+                .consensus_state()
+                .unwrap()
+                .next_epoch_data
+                .lock_checkpoint;
+    } else {
+        return false;
```
Comment (creativcoder) [inline consensus/src/checkpoint.rs:75]: we can remove `return` from both branches

Comment (polarbeargo) [reply consensus/src/checkpoint.rs:75]: If remove return build will fail.
Screenshot (305)

File: consensus/src/checkpoint.rs [hunk 6]
```diff
@@ -0,0 +1,163 @@
+// Copyright 2020 ChainSafe Systems
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::common::{Chain, ProtocolStateChain};
+use mina_crypto::base58::Base58Encodable;
+use mina_crypto::hash::{EpochSeed, StateHash};
+use mina_rs_base::protocol_state::ProtocolState;
+#[allow(dead_code)]
```
Comment (creativcoder) [inline consensus/src/checkpoint.rs:8]: I don't think we need this. This is being used below right?

Comment (polarbeargo) [reply consensus/src/checkpoint.rs:8]: We use this in the init_checkpoints()

Comment (creativcoder) [reply consensus/src/checkpoint.rs:8]: then this attribute should not be there, because the lint says dead_code

Comment (polarbeargo) [reply consensus/src/checkpoint.rs:8]: If we don't use this tag these Clippy warnings will come out again

Comment (polarbeargo) [reply consensus/src/checkpoint.rs:8]: It actually used at [here ](https://github.com/ChainSafe/mina-rs/blob/a69a2d09029df9ef739a4d4a9e77d8b7b014b39b/consensus/src/checkpoint.rs#L62)but Clippy think it's never used [Github Action](https://github.com/ChainSafe/mina-rs/runs/4284862218?check_suite_focus=true#step:9:20)

Comment (creativcoder) [reply consensus/src/checkpoint.rs:8]: ok, let's tackle this later.

**Review by creativcoder — APPROVED** [review]

**Comment by creativcoder** [comment]
LGTM, good for merge after a squash

File: crypto/src/hash/types.rs [hunk 7]
```diff
@@ -18,7 +18,7 @@ use wire_type::WireType;
 
 pub(crate) type HashBytes = Box<[u8]>;
 
-#[derive(Clone, Default, Debug, PartialEq, Serialize, Deserialize)]
+#[derive(Copy, Clone, Default, Debug, PartialEq, Serialize, Deserialize)]
```
Comment (creativcoder) [inline crypto/src/hash/types.rs:21]: Can we not use Copy here. Deriving Copy is implicit here. Instead we can call `clone()` where we need them, or just take ref of the BaseHash

Comment (polarbeargo) [reply crypto/src/hash/types.rs:21]: We have to use Copy to resolve this error "[ Use of moved value.](https://github.com/ChainSafe/mina-rs/pull/80/commits/68772460e890148d6d1630eac2248255bb480ac2)"

File: consensus/src/checkpoint.rs [hunk 8]
```diff
@@ -0,0 +1,163 @@
...
+        .lock_checkpoint = StateHash::default();
+    genesis
+        .body
+        .consensus_state
+        .staking_epoch_data
+        .epoch_length
+        .0 = 1;
+    genesis.body.consensus_state.next_epoch_data.seed =
+        Base58Encodable::from_base58("2vaRh7FQ5wSzmpFReF9gcRKjv48CcJvHs25aqb3SSZiPgHQBy5Dt")
+            .unwrap();
```
Comment (hanabi1224) [inline consensus/src/checkpoint.rs:32]: We should not call `.unwrap` in non-test code, either return Result or use expect

File: consensus/src/checkpoint.rs [hunk 9]
```diff
@@ -0,0 +1,163 @@
...
+    genesis
+        .body
+        .consensus_state
+        .next_epoch_data
+        .start_checkpoint = StateHash::default();
+    genesis.body.consensus_state.next_epoch_data.lock_checkpoint =
+        Base58Encodable::from_base58("3NLoKn22eMnyQ7rxh5pxB6vBA3XhSAhhrf7akdqS6HbAKD14Dh1d")
+            .unwrap();
```
Comment (hanabi1224) [inline consensus/src/checkpoint.rs:40]: ditto

File: consensus/src/checkpoint.rs [hunk 10]
```diff
@@ -0,0 +1,163 @@
...
+        .staking_epoch_data
+        .lock_checkpoint = StateHash::default();
+    genesis
+        .body
+        .consensus_state
+        .staking_epoch_data
+        .epoch_length
+        .0 = 1;
+    genesis.body.consensus_state.next_epoch_data.seed =
+        Base58Encodable::from_base58("2vaRh7FQ5wSzmpFReF9gcRKjv48CcJvHs25aqb3SSZiPgHQBy5Dt")
```
Comment (hanabi1224) [inline consensus/src/checkpoint.rs:31]: if this is a static array, use &'static T in the struct instead

File: consensus/src/checkpoint.rs [hunk 11]
```diff
@@ -0,0 +1,163 @@
...
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::common::{Chain, ProtocolStateChain};
+use mina_crypto::base58::Base58Encodable;
+use mina_crypto::hash::{EpochSeed, StateHash};
+use mina_rs_base::protocol_state::ProtocolState;
+#[allow(dead_code)]
+const SLOTS_PER_EPOCH: u32 = 7140;
+
+#[allow(dead_code)]
```
Comment (hanabi1224) [inline consensus/src/checkpoint.rs:11]: why dead code instead of making the funs pub?

File: consensus/src/checkpoint.rs [hunk 12]
```diff
@@ -0,0 +1,163 @@
...
+            .staking_epoch_data
+            .lock_checkpoint
+            == c1
+                .consensus_state()
+                .unwrap()
+                .staking_epoch_data
+                .lock_checkpoint;
+    }
+
+    #[allow(clippy::erasing_op)]
```
Comment (hanabi1224) [inline consensus/src/checkpoint.rs:57]: why is this needed?

File: consensus/src/checkpoint.rs [hunk 13]
```diff
@@ -0,0 +1,161 @@
...
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::common::{Chain, ProtocolStateChain};
+use mina_crypto::base58::Base58Encodable;
+use mina_crypto::hash::{EpochSeed, StateHash};
+use mina_rs_base::protocol_state::ProtocolState;
+
+pub const SLOTS_PER_EPOCH: u32 = 7140;
+
+pub fn init_checkpoints(genesis: &mut ProtocolState) {
```
Comment (hanabi1224) [inline consensus/src/checkpoint.rs:11]: why not impl init_checkpoints  and is_short_range  for ProtocolState and ProtocolStateChain respectively

Comment (polarbeargo) [reply consensus/src/checkpoint.rs:11]: They're not pure functions as params are mutable

File: consensus/src/lib.rs [hunk 14]
```diff
@@ -1,10 +1,5 @@
 // Copyright 2020 ChainSafe Systems
 // SPDX-License-Identifier: Apache-2.0
 
-#![deny(rustdoc::all)]
```
Comment (hanabi1224) [inline consensus/src/lib.rs:4]: plz revert this bad merge

File: consensus/src/checkpoint.rs [hunk 15]
```diff
@@ -0,0 +1,158 @@
...
+        Base58Encodable::from_base58("3NLoKn22eMnyQ7rxh5pxB6vBA3XhSAhhrf7akdqS6HbAKD14Dh1d")
+            .map_err(|_| ConsensusErrTyp::ConsensusInitFail)?;
+    genesis.body.consensus_state.next_epoch_data.epoch_length.0 = 2;
+    Ok(())
+}
+
+pub fn is_short_range(
+    c0: &ProtocolStateChain,
+    c1: &ProtocolStateChain,
+) -> Result<bool, ConsensusErrTyp> {
```
Comment (hanabi1224) [inline consensus/src/checkpoint.rs:51]: OK for now, but if u look at [ocaml code](https://github.com/MinaProtocol/mina/blob/9dd0c963e87f80c81006c6ecda34efbf6543d2a3/src/lib/consensus/proof_of_stake.ml#L2919), I think we can safely return false for the err condition

Comment (polarbeargo) [reply consensus/src/checkpoint.rs:51]: For not use unwrap() https://github.com/ChainSafe/mina-rs/pull/80/commits/c15385f425f4344052ef57f3e27bcee1a5f621c2#

File: consensus/src/checkpoint.rs [hunk 16]
```diff
@@ -0,0 +1,159 @@
...
+    let s0_lock_checkpoint = &s0.staking_epoch_data.lock_checkpoint;
+    let s1_lock_checkpoint = &s1.staking_epoch_data.lock_checkpoint;
+    let s1_next_epoch_lock_checkpoint = &s1.next_epoch_data.lock_checkpoint;
+
+    if s0.epoch_count == s1.epoch_count {
+        return Ok(s0_lock_checkpoint == s1_lock_checkpoint)
+    }
+
+    if s0.epoch_count.0 == s1.epoch_count.0 + 1
+        && Chain::epoch_slot(c1) >= Some(SLOTS_PER_EPOCH * 2 / 3)
```
Comment (creativcoder) [inline consensus/src/checkpoint.rs:67]: instead of `Chain::epoch_slot`, we can do `c1.epoch_slot()`

**Review by hanabi1224 — APPROVED** [review]

File: consensus/src/checkpoint.rs (hunk 16 above)
Comment (polarbeargo) [reply consensus/src/checkpoint.rs]: Thank you @creativcoder 😄
